trace_decoder/src/processed_block_trace_mpt.rs

```diff
@@ -0,0 +1,400 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::iter::once;
+
+use ethereum_types::{Address, H256, U256};
+use evm_arithmetization_mpt::generation::mpt::{AccountRlp, LegacyReceiptRlp};
+use evm_arithmetization_mpt::GenerationInputs;
+use mpt_trie::nibbles::Nibbles;
+use mpt_trie::partial_trie::PartialTrie;
+
+use crate::code_hash_resolver::CodeHashResolver;
+use crate::compact::compact_prestate_processing::{
+    MptPartialTriePreImages, ProcessedCompactOutput,
+};
+use crate::decoding_mpt::TxnMetaState;
+use crate::processed_block_trace::ProcessedBlockTrace;
+use crate::protocol_processing::{
+    process_mpt_block_trace_trie_pre_images, TraceProtocolDecodingResult,
+};
+use crate::trace_protocol::{AtomicUnitInfo, BlockTrace, ContractCodeUsage, TxnInfo};
+use crate::types::{
+    CodeHash, CodeHashResolveFunc, HashedAccountAddr, HashedNodeAddr, HashedStorageAddrNibbles,
+    OtherBlockData, TrieRootHash, EMPTY_CODE_HASH, EMPTY_TRIE_HASH,
+};
+use crate::utils::{
+    hash, print_value_and_hash_nodes_of_storage_trie, print_value_and_hash_nodes_of_trie,
+};
+
+pub(crate) type MptProcessedBlockTrace = ProcessedBlockTrace<ProcedBlockTraceMptSpec>;
+
+pub(crate) type StorageAccess = Vec<HashedStorageAddrNibbles>;
+pub(crate) type StorageWrite = Vec<(HashedStorageAddrNibbles, Vec<u8>)>;
+
+#[derive(Debug)]
+pub(crate) struct ProcedBlockTraceMptSpec {
+    pub(crate) tries: MptPartialTriePreImages,
+    pub(crate) sect_info: ProcessedSectionInfo,
+}
+
+impl BlockTrace {
+    /// Processes and returns the [GenerationInputs] for all transactions in the
+    /// block.
+    pub fn into_proof_gen_mpt_ir<F>(
+        self,
+        c_resolve: &dyn CodeHashResolveFunc,
+        other_data: OtherBlockData,
+    ) -> TraceProtocolDecodingResult<Vec<GenerationInputs>>
+    where
+        F: CodeHashResolveFunc,
+    {
+        let processed_block_trace =
+            self.into_mpt_processed_block_trace(c_resolve, other_data.b_data.withdrawals.clone())?;
+
+        let res = processed_block_trace.into_proof_gen_ir(other_data)?;
+
+        Ok(res)
+    }
+
+    fn into_mpt_processed_block_trace(
+        self,
+        c_resolve: &dyn CodeHashResolveFunc,
```

Comment:
I'm open to both dynamic dispatch (`c_resolve: &dyn CodeHashResolver`) or static dispatch (`c_resolve: T`) here, but I think static dispatch is the slightly better choice here, since the client is really only going to ever create one impl of this.

Reply:
```
error[E0782]: trait objects must include the `dyn` keyword
  --> trace_decoder/src/processed_block_trace_mpt.rs:44:21
   |
44 |         c_resolve: &CodeHashResolveFunc,
   |                     ^^^^^^^^^^^^^^^^^^^
   |
help: add `dyn` keyword before this trait
   |
44 |         c_resolve: &dyn CodeHashResolveFunc,
   |                     +++
   ```

Reply:
I am unable to use static dispatch here
It is asking me to use dynamic only